app/models/manageiq/providers/ibm_cloud/vpc/cloud_manager/vm.rb

```diff
@@ -21,4 +21,89 @@ def self.calculate_power_state(raw_power_state)
   def self.display_name(number = 1)
     n_('Instance (IBM)', 'Instances (IBM)', number)
   end
+
+  # Used in with_provider_object to scope SDK to this instance.
+  def provider_object(vpc)
+    vpc.instances.instance(ems_ref)
+  end
+
+  # Send a start action to IBM Cloud. Wait for state to change to started, then update the raw_power_state.
+  def raw_start
+    with_provider_object do |instance|
+      instance.actions.start
+      instance.wait_for! do
+        sdk_update_status(instance)
+        instance.started?
+      end
+    end
+  rescue => e
+    $ibm_cloud_log.error(e.to_s)
+    $ibm_cloud_log.log_backtrace(e)
+    raise
+  end
+
+  # IBM Cloud does not support suspend. Hiding it from UI.
+  supports_not :suspend
+
+  # Send a stop action to IBM Cloud. Wait for state to change to stopped, then update the raw_power_state.
+  def raw_stop
+    with_provider_object do |instance|
+      instance.actions.stop
+      instance.wait_for! do
+        sdk_update_status(instance)
+        instance.stopped?
+      end
+    end
+  rescue => e
+    $ibm_cloud_log.error(e.to_s)
+    $ibm_cloud_log.log_backtrace(e)
+    raise
+  end
+
+  # IBM Cloud does not support pause. Using stop since can't hide it in UI.
+  def raw_pause
+    raw_stop
+  end
+
+  # Show reboot in the instance menu when on.
+  supports :reboot_guest do
+    unsupported_reason_add(:reboot_guest, _('The VM is not powered on')) unless current_state == 'on'
+  end
+
+  # Gracefully reboot the quest.
+  # @param force [Boolean] Ungracefully reboot VM.
+  def raw_reboot_guest(force: false)
+    with_provider_object do |instance|
+      instance.actions.reboot(force)
+      sleep 5 # Sleep for 5 seconds to allow for reboot sequence to start.
```

Comment:
This is unfortunate but I get what you're trying to avoid here, is there any chance that starting a graceful shutdown might take longer than 5 seconds?

Reply:
Right, I agree. I had a wait_for transitional? but the VMs I'm testing again reboot way faster than the wait_for loop allow even with a 0.5 second sleep. It was for this purpose that I decided now was the time to implement the block. However when I tried testing my code the vm got stuck in the transitional loop. I guess we could do a wait for transitional in a begin with a 20 second timeout and just ignore the exception on timeout.

Reply:
Either we do a begin and ignore the exception, which is bad style, or we open up the SDK again and implement the tuple wait_for. It wouldn't be hard.

```
def wait_for(sleep_time: 5, timeout: 600, &block)
  @logger.info("Starting wait for instance #{id}. Starts in state #{status}.")
  loop do
    refresh
    return (false, "VM #{id} is in a failed state.") if failed?
    break if block.call(self)

    timeout = sleep_counter(sleep_time, timeout)
    return (false, "Time out while waiting #{id} to be stable.") if timeout <= 0
  end
  @logger.info("Finished wait for instance #{id}. Ends in state #{status}.")
  return (true, 'ok')
end

def wait_for!(sleep_time: 5, timeout: 600, &block)
  status, msg = wait_for(sleep_time: sleep_time, timeout: timeout, &block)
  return if status
  raise msg
end
```

Reply:
As for your initial question. The 5 second wait is to allow the VM to go into starting, stopping or some other state. The wait_for without a sleep will just see the started vm and return, which is a null operation.

Reply:
Right I assumed it was only in the reboot operation because you're waiting for the VM to be started and if you check too quickly the vm will still be running and hasn't started to reboot yet.

Reply:
Worst case if we return before a VM has started rebooting, we still issued the API call successfully.  I'd have to check if any other providers wait for the Vm to fully restart before returning.  I know VMware doesn't because it isn't an operation that returns a task.

Returning before the operation has completed is fine as long as it was initiated successfully

Reply:
Ok for the time being we can keep the sleep. I need to get some specs written for this because I keep testing in the UI, changing something and forgetting to test again.